hbase-server/src/test/java/org/apache/hadoop/hbase/client/AbstractTestUpdateConfiguration.java

```diff
@@ -0,0 +1,103 @@
+  private static Path overwriteConfigFileUnderTestDataDir;
+  private static Path backupConfigFileUnderTestDataDir;
+
+  protected static void setUpConfigurationFiles(final HBaseTestingUtility testUtil)
+    throws Exception {
+    // Before this change, the test will update hbase-site.xml under target/test-classes and
+    // trigger a config reload. Since target/test-classes/hbase-site.xml is being used by
+    // other testing cases at the same time, this update will break other testing cases so it will
+    // be flakey in nature.
+    // To avoid this, the change is to make target/test-classes/hbase-site.xml immutable. A new
+    // hbase-site.xml will be created under its test data directory, i.e,
+    // hbase-server/target/test-data/UUID, this new file will be added as a resource for the
+    // config, new update will be applied to this new file and only visible to this specific test
+    // case. The target/test-classes/hbase-site.xml will not be changed during the test.
+
+    String absoluteDataPath = testUtil.getDataTestDir().toString();
+
+    // Create test-data directories.
+    Files.createDirectories(Paths.get(absoluteDataPath));
+
+    // Copy hbase-site.xml from target/test-class to target/test-data/UUID directory.
+    Path configFile = Paths.get("target", "test-classes", SERVER_CONFIG);
+    configFileUnderTestDataDir = Paths.get(absoluteDataPath, SERVER_CONFIG);
+    Files.copy(configFile, configFileUnderTestDataDir);
+
+    // Copy overwrite config file overwrite-hbase-site.xml from target/test-class to
+    // target/test-data/UUID directory.
+    Path overwriteConfigFile = Paths.get("target", "test-classes",
+      OVERWRITE_SERVER_CONFIG);
+    overwriteConfigFileUnderTestDataDir = Paths.get(absoluteDataPath, OVERWRITE_SERVER_CONFIG);
+    Files.copy(overwriteConfigFile, overwriteConfigFileUnderTestDataDir);
+
+    backupConfigFileUnderTestDataDir = Paths.get(absoluteDataPath, BACKUP_SERVER_CONFIG);
+
+    // Add the new custom config file to Configuration
+    testUtil.getConfiguration().addResource(testUtil.getDataTestDir(SERVER_CONFIG));
+  }
+
+  protected static void addResourceToRegionServerConfiguration(final HBaseTestingUtility testUtil) {
+    // When RegionServer is created in MiniHBaseCluster, it uses HBaseConfiguration.create(conf) of
+    // the master Configuration. The create() just copies config params over, it does not do
+    // a clone for a historic reason. Properties such as resources are lost during this process.
+    // Exposing a new method in HBaseConfiguration causes confusion. Instead, the new hbase-site.xml
+    // under test-data directory is added to RegionServer's configuration as a workaround.
+    for (RegionServerThread rsThread : testUtil.getMiniHBaseCluster().getRegionServerThreads()) {
+      rsThread.getRegionServer().getConfiguration().addResource(
+        testUtil.getDataTestDir(SERVER_CONFIG));
+    }
+  }
+
+  protected void replaceHBaseSiteXML() throws IOException {
```

Comment:
since these methods are part of an interface consumed by children, some javadoc would be nice here. Something like

> replace the site.xml file under this test's data directory with the content of the override file. stashes the current existing file so that it can be restored using {@link #restoreHBaseSiteXML()}

and 

> Restores the site.xml that was stashed by a previous call to {@link #replaceHBaseSiteXML()}.

Reply:
Let me address it.